### PR TITLE
add can

### PR DESCRIPTION
**Pull request by xoviat**

### CODE REVIEW

**Comment by xoviat** [comment]
cc @Dirbaio

**Comment by xoviat** [comment]
dependency conflict: stm32-rs/bxcan#18

File: embassy-stm32f4/Cargo.toml [hunk 1]
```diff
@@ -39,3 +39,5 @@ cortex-m = "0.7.1"
 embedded-hal    = { version = "0.2.4" }
 embedded-dma    = { version = "0.1.2" }
 stm32f4xx-hal  = { version = "0.8.3", features = ["rt"], git = "https://github.com/stm32-rs/stm32f4xx-hal.git"}
+bxcan = {version = "0.4.0", git= "https://github.com/jonas-schievink/bxcan.git", branch = "defmt02" }
```
Comment (marcelbuesing) [inline embassy-stm32f4/Cargo.toml:42]: Looks like this branch no longer exists but changes have been merged to
https://github.com/stm32-rs/bxcan

```toml
bxcan = { version = "0.4.0", git= "https://github.com/stm32-rs/bxcan.git", branch = "master" }
```

File: embassy-stm32f4/src/can.rs [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+use nb;
+use nb::block;
+
+use crate::hal::prelude::*;
+use crate::interrupt;
+
+/// Interface to the Serial peripheral
+pub struct Can<T: bxcan::Instance> {
+    can: bxcan::Can<T>,
+    tx_int: interrupt::CAN1_TX,
```
Comment (Dirbaio) [inline embassy-stm32f4/src/can.rs:22]: the interrupts may be wrong if user doesn't use CAN1? 

How does `Instance` work in bxcan? the user has to implement it themselves for their chip?  Maybe we can provide the impls?

**Comment by Dirbaio** [comment]
Looks great! I'm surprised so very little code is needed, nice!

Can you add an example?

File: embassy-stm32f4/src/can.rs [hunk 3]
```diff
@@ -78,3 +83,30 @@ impl<T: bxcan::Instance> Can<T> {
         }
     }
 }
+
+mod private {
+    pub trait Sealed {}
+}
+
+pub trait WithInterrupt: bxcan::Instance + private::Sealed {
```
Comment (Dirbaio) [inline embassy-stm32f4/src/can.rs:87]: This could simply be named `Instance`.

File: embassy-stm32f4/src/can.rs [hunk 4]
```diff
@@ -78,3 +83,30 @@ impl<T: bxcan::Instance> Can<T> {
...
+    type TInstance;
+    type RInstance;
```
Comment (Dirbaio) [inline embassy-stm32f4/src/can.rs:89]: add ` : Interrupt` trait bounds here

Comment (Dirbaio) [reply embassy-stm32f4/src/can.rs:89]: Maybe name TxInterrupt/RxInterrupt? They're not peripheral instances, they're interrupts

File: embassy-stm32f4/src/can.rs [hunk 5]
```diff
@@ -17,13 +17,18 @@ use crate::hal::prelude::*;
 use crate::interrupt;
 
 /// Interface to the Serial peripheral
-pub struct Can<T: bxcan::Instance> {
+pub struct Can<T: bxcan::Instance, TINT: Interrupt, RINT: Interrupt> {
```
Comment (Dirbaio) [inline embassy-stm32f4/src/can.rs:20]: There's no need for having 3 generic params anymore. Makes the type simpler:

```rust
pub struct Can<T: Instance> {
    can: bxcan::Can<T>,
    tx_int: T::TxInterrupt,
    rx_int: T::RxInterrupt,
}
```

**Review by Dirbaio — CHANGES_REQUESTED** [review]

**Comment by xoviat** [comment]
This needs to go back to draft, because the defmt update caused problems.

**Comment by xoviat** [comment]
pending stm32-rs/bxcan#21

**Pull request merged**